elastic/beats
# [Kubernetes - Metricbeat] Providing config option to disable Kubeadm config api requests

- Enhancement

~**NOTE:** Dont merge until https://github.com/elastic/elastic-agent-autodiscover/pull/98 is also merged~

The elastic-agent-autodoscover library updated to [v0.8.2](https://github.com/elastic/beats/pull/40086/files#diff-3295df7234525439d778f1b282d146a4f1ff6b415248aaac074e8042d9f42d63R559)

## Proposed commit message

**WHAT**: Provide a configuration option to disable `/api/v1/namespaces/kube-system/configmaps/kubeadm-config` . In more details user will disable the api requests to the modules he configures by using configuration `use_kubeadm: false`

**WHY**: https://github.com/elastic/enhancements/issues/21172

## Checklist


## How to test this PR locally

- Install a kubernetes cluster with audit-logs enabled
- Create a local elastic stack with `elastic-package stack up -d -v --version=8.15.0-SNAPSHOT`
- Update the go.mod file of this repo and point to the elastic-agent-autodiscovery folder, where related changes should also exist. Your go.mod file should have the following entry: 
```bash
replace github.com/elastic/elastic-agent-autodiscover => /Users/andreasgkizas/elastic/elastic-agent-autodiscover
```
- Run `go mod tidy`
- Install Filebeat and collect audit logs
- Follow instructions of this [page](https://github.com/elastic/beats/blob/main/metricbeat/module/kubernetes/_meta/test/docs/README.md) to build the metricbeat binary
```bash
 GOOS=linux GOARCH=arm64 go build
 kubectl apply -f module/kubernetes/_meta/test/docs/01_playground/metricbeat.yaml
 kubectl cp metricbeat `kubectl get pod -n kube-system -l k8s-app=metricbeat -o jsonpath='{.items[].metadata.name}'`:/usr/share/metricbeat/ -n kube-system\n
 kubectl exec `kubectl get pod -n kube-system -l k8s-app=metricbeat -o jsonpath='{.items[].metadata.name}'` -n kube-system -- bash -c "metricbeat -e -c /etc/metricbeat.yml"
```
- Use `use_kubeadm: false` flag in your enabled modules
- Open your local kibana cluster, navigate to Discovery and search inside kibana audit-logs for relevant api requests
<img width="1717" alt="Screenshot 2024-07-03 at 12 06 49 PM" src="https://github.com/elastic/beats/assets/1685415/b34110fb-4f0b-4eae-b37c-603074833132">

## Related issues

-  Relates #https://github.com/elastic/enhancements/issues/21172
- Relates #https://github.com/elastic/elastic-agent-autodiscover/pull/98

## Screenshots

### With 8.15.0 Image no changes:

<img width="1717" alt="Screenshot 2024-07-03 at 12 06 49 PM" src="https://github.com/elastic/beats/assets/1685415/7039bc0c-9817-485c-b80a-57e1303d56b1">

We enable all metricsets (included events) and we have 85 API requests

### With This repo code but the flag is `disable_kubeadm: false`

<img width="1717" alt="Screenshot 2024-07-03 at 12 34 26 PM" src="https://github.com/elastic/beats/assets/1685415/c474d518-e0bb-4179-93f9-ec38dbfa61c5">

We enable all metricsets (included events) and we still have 85 API requests


### With This repo code but the flag is disable_kubeadm: true`

<img width="1717" alt="Screenshot 2024-07-03 at 1 08 16 PM" src="https://github.com/elastic/beats/assets/1685415/096a8d09-3b95-4444-9fd8-2eeefeb0f222">

We have no API requests after 10:04 that metrics collection established

```json
{"log.level":"info","@timestamp":"2024-07-03T10:04:20.520Z","log.logger":"publisher_pipeline_output","log.origin":{"function":"github.com/elastic/beats/v7/libbeat/publisher/pipeline.(*netClientWorker).run","file.name":"pipeline/client_worker.go","file.line":145},"message":"Connection to backoff(elasticsearch(https://elasticsearch:9200)) established","service.name":"metricbeat","ecs.version":"1.6.0"}
```
